Add workqueue example

rmohr opened the pull request:
Demonstrates how to compose a controller out of cache.Controller, cache.Indexer and a workqueue.

Tested on kubernetes 1.4.5.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/client-go/65)
<!-- Reviewable:end -->

rmohr commented:
cc @caesarxuchao @wfarr

lavalamp commented on examples/workqueue/main.go:
```diff
@@ -0,0 +1,197 @@
+			return false
+		}
+		// Tell the queue that we are done with processing this key. This unblocks the key for other workers
+		// This allows save parallel processing because two pods with the same key are never processed in
```
s/save/safe/

lavalamp commented on examples/workqueue/main.go:
```diff
@@ -0,0 +1,197 @@
+			} else {
+				queue.Forget(key)
+			}
+			glog.Fatalf("Invalid key %s", key.(string))
```
this line seems wrong

lavalamp commented on examples/workqueue/main.go:
```diff
@@ -0,0 +1,197 @@
+
+		if !exists {
+			// Below we will warm up our cache with a Pod, so that we will see a delete for one pod
+			fmt.Printf("Pod %s does not exist anymore\n", key.(string))
```
Use glog for log messages (or the built-in log package).

lavalamp commented on examples/workqueue/main.go:
```diff
@@ -0,0 +1,197 @@
+	// Wait forever
+	for {
+		time.Sleep(60000 * time.Millisecond)
+	}
```
select {} is a more elegant way to block forever

lavalamp commented on examples/workqueue/main.go:
```diff
@@ -0,0 +1,197 @@
+type ControllerFunc func(cache.Indexer, workqueue.RateLimitingInterface) bool
+
+func (c *Controller) Run(threadiness int, stopCh chan struct{}) {
+	defer NewPanicCatcher()
```
This line isn't doing what you think it's doing :)

lavalamp commented on examples/workqueue/main.go:
```diff
@@ -0,0 +1,197 @@
+	// Create the connection
+	config, err := clientcmd.BuildConfigFromFlags(master, kubeconfig)
+	if err != nil {
+		panic(err.Error())
```
glog.Fatalf

lavalamp commented on examples/workqueue/main.go:
```diff
@@ -0,0 +1,197 @@
+import (
+	"flag"
+	"fmt"
+	"github.com/golang/glog"
```
add blank lines so that there's three sections: builtin packages, k8s.io packages, 3rd party dependencies

lavalamp commented:
thanks for the example!

rmohr commented:
Thx for the review. Should have addressed all issues, but I am still not sure what the best practice is when we can't retrieve an item from the store in the process loop.

---

Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[examples/workqueue/main.go, line 22 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/client-go/65#-KaHeuPkn_qB0EUXBTLF-r1-22:-KaM8Z_3IwvyqxfzYn-7:b-896fix) ([raw file](https://github.com/kubernetes/client-go/blob/dc7da3f40561c48f8769727631fc4abb49c045a0/examples/workqueue/main.go#L22)):*
<details><summary><i>Previously, lavalamp (Daniel Smith) wrote…</i></summary><blockquote>

add blank lines so that there's three sections: builtin packages, k8s.io packages, 3rd party dependencies
</blockquote></details>

Done.

---

*[examples/workqueue/main.go, line 64 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/client-go/65#-KaHeuPkn_qB0EUXBTLF-r1-64:-KaM8gGanS5hRDxcLEVz:b-896fix) ([raw file](https://github.com/kubernetes/client-go/blob/dc7da3f40561c48f8769727631fc4abb49c045a0/examples/workqueue/main.go#L64)):*
<details><summary><i>Previously, lavalamp (Daniel Smith) wrote…</i></summary><blockquote>

This line isn't doing what you think it's doing :)
</blockquote></details>

Done.

---

*[examples/workqueue/main.go, line 89 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/client-go/65#-KaHeuPkn_qB0EUXBTLF-r1-89:-KaM8h7CYpoiQRR9JLbh:b-896fix) ([raw file](https://github.com/kubernetes/client-go/blob/dc7da3f40561c48f8769727631fc4abb49c045a0/examples/workqueue/main.go#L89)):*
<details><summary><i>Previously, lavalamp (Daniel Smith) wrote…</i></summary><blockquote>

glog.Fatalf
</blockquote></details>

Done.

---

*[examples/workqueue/main.go, line 138 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/client-go/65#-KaHeuPkn_qB0EUXBTLF-r1-138:-KaM8i92IY45vBRTb_ky:b-896fix) ([raw file](https://github.com/kubernetes/client-go/blob/dc7da3f40561c48f8769727631fc4abb49c045a0/examples/workqueue/main.go#L138)):*
<details><summary><i>Previously, lavalamp (Daniel Smith) wrote…</i></summary><blockquote>

s/save/safe/
</blockquote></details>

Done.

---

*[examples/workqueue/main.go, line 156 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/client-go/65#-KaHeuPkn_qB0EUXBTLF-r1-156:-KaM8jgzycK3yfPem5co:b8ztmyv) ([raw file](https://github.com/kubernetes/client-go/blob/dc7da3f40561c48f8769727631fc4abb49c045a0/examples/workqueue/main.go#L156)):*
<details><summary><i>Previously, lavalamp (Daniel Smith) wrote…</i></summary><blockquote>

this line seems wrong
</blockquote></details>

Should be more correct now. Still don't know if this is a good practice what I do here ...

---

*[examples/workqueue/main.go, line 162 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/client-go/65#-KaHeuPkn_qB0EUXBTLF-r1-162:-KaM8pmygPEzqHBY9e7_:b-896fix) ([raw file](https://github.com/kubernetes/client-go/blob/dc7da3f40561c48f8769727631fc4abb49c045a0/examples/workqueue/main.go#L162)):*
<details><summary><i>Previously, lavalamp (Daniel Smith) wrote…</i></summary><blockquote>

Use glog for log messages (or the built-in log package).
</blockquote></details>

Done.

---

*[examples/workqueue/main.go, line 188 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/client-go/65#-KaHeuPkn_qB0EUXBTLF-r1-188:-KaM8vKex7Iur5AS49Sn:b-896fix) ([raw file](https://github.com/kubernetes/client-go/blob/dc7da3f40561c48f8769727631fc4abb49c045a0/examples/workqueue/main.go#L188)):*
<details><summary><i>Previously, lavalamp (Daniel Smith) wrote…</i></summary><blockquote>

select {} is a more elegant way to block forever
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/client-go/65#-:-KaM8zH-isjxp50dqrnR:brb7dok)*
<!-- Sent from Reviewable.io -->

ash2k commented on examples/workqueue/main.go:
```diff
@@ -0,0 +1,194 @@
+func init() {
+	flag.StringVar(&kubeconfig, "kubeconfig", "", "absolute path to the kubeconfig file")
+	flag.StringVar(&master, "master", "", "master url")
+}
```
Why not put flags and global variables into `main()`? Global state hurts code reusability/etc.

ash2k commented on examples/workqueue/main.go:
```diff
@@ -0,0 +1,194 @@
+
+func (c *Controller) Run(threadiness int, stopCh chan struct{}) {
+	// Don't crash the loop on a panic
+	defer catchPanic()
```
IMHO this is not relevant for this example. Also, again IMHO, it is not a great design choice :) Panic means something is completely wrong and there is no way to handle it. Why catch it?

rmohr commented:
I signed it

rmohr commented:
@k8s-ci-robot I signed it

rmohr commented:
@k8s-ci-robot  I signed it

rmohr commented:
@caesarxuchao @lavalamp anything more you want me to change?

caesarxuchao commented:
@mbohlool it would be great if you can also take a look at the pull.

mbohlool commented:
@rmohr @caesarxuchao Should this PR sent against main repo vs client-go repo?

caesarxuchao commented:
The main repo please. It will be copied over to client-go.

rmohr commented:
Ok, will do a PR against the main repo :+1:

mbohlool commented:
@rmohr Thanks. Let me know when you send that PR, and I will review it with priority.

ymruan commented:
workqueue is not in client-go 2.0, which kind of queue should I use?

caesarxuchao commented:
@ymruan is it possible to use the workqueue in k8s.io/kubernetes release-1.5? I guess it will work.
